### DISCUSSION TITLE
Highlight match of pair left of the cursor in insert mode

### DISCUSSION THREAD

**Discussion by TheZoq2** [helix-editor/helix#8548] [Ideas]
In the editors I'm used to (neovim, vscode), when in insert mode, the pair highlighting highlights the sibling of the pair to the left of the cursor when there is one, but helix seems to do the opposite.

This is probably easier to demonstrate with an example (`|` is the cursor)
```
((()|
  ^^
```
or
```
((())|
 ^  ^
```
Currently, helix seems to not highlight any of the parens in these situations. I find this very annoying, because I often use the pair highlighting to figure out if I need to add more parens. Here I have to guess or exit insert mode to check.

Of course, this becomes a bit ambiguous with matched parens, which neovim highlights like this,
```
((()|))
  ^^
```
but helix like
```
((()|))
 ^   ^
```
 but for that to be intuitive, the highlighting needs to highlight both pairs (#8489)

**Comment by pascalkuthe** [comment]
The issue is that you viel the cursors as a bar instead of a Block. Helix used Block cursors semantica, where (even in insertmode) the cursors always has width one. So we don't highlight the brackets to the kwft of the cursors or to the right of the cursors but the bracket **at** the cursors.

It would be weird to apply the matching brackets highlight anywhere but the cursors itself. I don't think we will go forward with that

**Reply by TheZoq2** [reply]
I do see the reason for it, but I'm also noticing that I have a much harder time getting my bracket matching correct with helix than I do in neovim.

Since there is a `editor.cursor-shape`, would you be open to making an adjustment in insert mode, if the cursor is set to `bar` in insert mode?

As an experiment I'm adding it myself to a branch, and the change seems fairly minimal

**Reply by TheZoq2** [reply]
Here is my proposed change https://github.com/helix-editor/helix/commit/31576fa15619e85c594cfc1402ed9cc014906a33 though because of #8489, this behaves a bit strangely in `(|)`

**Comment by jba** [comment]
> It would be weird to apply the matching brackets highlight anywhere but the cursors itself. 

Conceptually, it's odd. In practice, it's quite useful. When I type a close brace in insert mode, I want to see immediately which open brace I'm matching. I don't want to move left one character, confirm the delimiter, then move right and keep typing.

Emacs has this, as does kakoune (as [`show-matching -previous`](https://github.com/mawww/kakoune/blob/master/doc/pages/highlighters.asciidoc#convenient-highlighters)).

I think this should be a configurable option, defaulting to false if you like, and independent of cursor shape.
